File: sdk/flutter/lib/src/customer_info.dart
import 'gen/moth/billing/v1/billing.pb.dart' as pb;
import 'gen/moth/billing/v1/billing.pbenum.dart' as pbe;

/// Which app store a purchase or subscription belongs to.
enum MothStore {
  apple,
  google;

  pbe.Store get proto => switch (this) {
    MothStore.apple => pbe.Store.STORE_APPLE,
    MothStore.google => pbe.Store.STORE_GOOGLE,
  };

  static MothStore? fromProto(pbe.Store store) => switch (store) {
    pbe.Store.STORE_APPLE => MothStore.apple,
    pbe.Store.STORE_GOOGLE => MothStore.google,
    _ => null,
  };
}

/// Why an entitlement is active.
enum MothEntitlementSource {
  /// Granted by an active store subscription.
  store,

  /// Granted by an operator (promo/comp), independent of store state.
  grant,

  /// The built-in free tier (no active subscription or grant).
  none;

  static MothEntitlementSource fromProto(pbe.EntitlementSource source) =>
      switch (source) {
        pbe.EntitlementSource.ENTITLEMENT_SOURCE_STORE =>
          MothEntitlementSource.store,
        pbe.EntitlementSource.ENTITLEMENT_SOURCE_GRANT =>
          MothEntitlementSource.grant,
        _ => MothEntitlementSource.none,
      };
}

/// The store's renewal state, mapped to a small set common to Apple and
/// Google. active/trialing/inGracePeriod/inBillingRetry keep access;
/// paused/expired/revoked do not.
enum MothSubscriptionStatus {
  unspecified,
  active,
  trialing,
  inGracePeriod,
  inBillingRetry,
  paused,
  expired,
  revoked;

  /// Whether this status keeps the subscription's access.
  bool get isActive => switch (this) {
    MothSubscriptionStatus.active ||
    MothSubscriptionStatus.trialing ||
    MothSubscriptionStatus.inGracePeriod ||
    MothSubscriptionStatus.inBillingRetry => true,
    _ => false,
  };

  static MothSubscriptionStatus fromProto(pbe.SubscriptionStatus status) =>
      switch (status) {
        pbe.SubscriptionStatus.SUBSCRIPTION_STATUS_ACTIVE =>
          MothSubscriptionStatus.active,
        pbe.SubscriptionStatus.SUBSCRIPTION_STATUS_TRIALING =>
          MothSubscriptionStatus.trialing,
        pbe.SubscriptionStatus.SUBSCRIPTION_STATUS_IN_GRACE_PERIOD =>
          MothSubscriptionStatus.inGracePeriod,
        pbe.SubscriptionStatus.SUBSCRIPTION_STATUS_IN_BILLING_RETRY =>
          MothSubscriptionStatus.inBillingRetry,
        pbe.SubscriptionStatus.SUBSCRIPTION_STATUS_PAUSED =>
          MothSubscriptionStatus.paused,
        pbe.SubscriptionStatus.SUBSCRIPTION_STATUS_EXPIRED =>
          MothSubscriptionStatus.expired,
        pbe.SubscriptionStatus.SUBSCRIPTION_STATUS_REVOKED =>
          MothSubscriptionStatus.revoked,
        _ => MothSubscriptionStatus.unspecified,
      };
}

/// One active capability the user holds (e.g. `pro`), with its expiry and why
/// it is active. Apps gate on [identifier], never on a product id.
class MothEntitlement {
  const MothEntitlement({
    required this.identifier,
    this.expireTime,
    this.source = MothEntitlementSource.none,
    this.productIdentifier = '',
  });

  factory MothEntitlement.fromProto(pb.Entitlement proto) => MothEntitlement(
    identifier: proto.identifier,
    expireTime: proto.hasExpireTime()
        ? proto.expireTime.toDateTime().toUtc()
        : null,
    source: MothEntitlementSource.fromProto(proto.source),
    productIdentifier: proto.productIdentifier,
  );

  factory MothEntitlement.fromJson(Map<String, Object?> json) =>
      MothEntitlement(
        identifier: json['identifier'] as String? ?? '',
        expireTime: json['expireTime'] == null
            ? null
            : DateTime.parse(json['expireTime'] as String),
        source: MothEntitlementSource.values.firstWhere(
          (s) => s.name == json['source'],
          orElse: () => MothEntitlementSource.none,
        ),
        productIdentifier: json['productIdentifier'] as String? ?? '',
      );

  /// Stable identifier the app checks (e.g. `pro`).
  final String identifier;

  /// When the entitlement lapses; null for a non-expiring grant.
  final DateTime? expireTime;

  /// Why it is active (store subscription vs operator grant).
  final MothEntitlementSource source;

  /// The moth product identifier that granted it, when [source] is
  /// [MothEntitlementSource.store]; empty for grants.
  final String productIdentifier;

  Map<String, Object?> toJson() => {
    'identifier': identifier,
    if (expireTime != null) 'expireTime': expireTime!.toUtc().toIso8601String(),
    'source': source.name,
    'productIdentifier': productIdentifier,
  };

  @override
  bool operator ==(Object other) =>
      other is MothEntitlement &&
      other.identifier == identifier &&
      other.expireTime == expireTime &&
      other.source == source &&
      other.productIdentifier == productIdentifier;

  @override
  int get hashCode =>
      Object.hash(identifier, expireTime, source, productIdentifier);
}

/// One of the user's store subscriptions (may be inactive, for history /
/// paywall display).
class MothActiveSubscription {
  const MothActiveSubscription({
    required this.productIdentifier,
    required this.store,
    required this.status,
    this.currentPeriodEnd,
    this.autoRenew = false,
    this.isSandbox = false,
  });

  factory MothActiveSubscription.fromProto(pb.ActiveSubscription proto) =>
      MothActiveSubscription(
        productIdentifier: proto.productIdentifier,
        store: MothStore.fromProto(proto.store),
        status: MothSubscriptionStatus.fromProto(proto.status),
        currentPeriodEnd: proto.hasCurrentPeriodEnd()
            ? proto.currentPeriodEnd.toDateTime().toUtc()
            : null,
        autoRenew: proto.autoRenew,
        isSandbox: proto.isSandbox,
      );

  factory MothActiveSubscription.fromJson(Map<String, Object?> json) =>
      MothActiveSubscription(
        productIdentifier: json['productIdentifier'] as String? ?? '',
        store: MothStore.values
            .where((s) => s.name == json['store'])
            .firstOrNull,
        status: MothSubscriptionStatus.values.firstWhere(
          (s) => s.name == json['status'],
          orElse: () => MothSubscriptionStatus.unspecified,
        ),
        currentPeriodEnd: json['currentPeriodEnd'] == null
            ? null
            : DateTime.parse(json['currentPeriodEnd'] as String),
        autoRenew: json['autoRenew'] as bool? ?? false,
        isSandbox: json['isSandbox'] as bool? ?? false,
      );

  final String productIdentifier;

  /// The store this subscription lives on; null when the server reported an
  /// unspecified store.
  final MothStore? store;
  final MothSubscriptionStatus status;

  /// End of the current paid (or trial) period; the renewal date when
  /// [autoRenew] is true.
  final DateTime? currentPeriodEnd;
  final bool autoRenew;

  /// Whether this subscription is a sandbox/test purchase.
  final bool isSandbox;

  Map<String, Object?> toJson() => {
    'productIdentifier': productIdentifier,
    if (store != null) 'store': store!.name,
    'status': status.name,
    if (currentPeriodEnd != null)
      'currentPeriodEnd': currentPeriodEnd!.toUtc().toIso8601String(),
    'autoRenew': autoRenew,
    'isSandbox': isSandbox,
  };

  @override
  bool operator ==(Object other) =>
      other is MothActiveSubscription &&
      other.productIdentifier == productIdentifier &&
      other.store == store &&
      other.status == status &&
      other.currentPeriodEnd == currentPeriodEnd &&
      other.autoRenew == autoRenew &&
      other.isSandbox == isSandbox;

  @override
  int get hashCode => Object.hash(
    productIdentifier,
    store,
    status,
    currentPeriodEnd,
    autoRenew,
    isSandbox,
  );
}

/// The complete subscription picture for one user, from
/// `moth.billing.v1.GetCustomerInfo`.
///
/// A never-paid user, a free-tier user, and a user in a project with no
/// products all get a well-formed instance with empty [activeEntitlements]
/// (the built-in `none` tier) — never an error. Gate features with
/// [hasEntitlement]; never special-case "never paid".
class MothCustomerInfo {
  const MothCustomerInfo({
    this.activeEntitlements = const [],
    this.subscriptions = const [],
  });

  /// The valid, empty state: no entitlements, the free `none` tier.
  const MothCustomerInfo.free() : this();

  factory MothCustomerInfo.fromProto(pb.CustomerInfo proto) => MothCustomerInfo(
    activeEntitlements: proto.activeEntitlements
        .map(MothEntitlement.fromProto)
        .toList(growable: false),
    subscriptions: proto.subscriptions
        .map(MothActiveSubscription.fromProto)
        .toList(growable: false),
  );

  factory MothCustomerInfo.fromJson(Map<String, Object?> json) =>
      MothCustomerInfo(
        activeEntitlements:
            (json['activeEntitlements'] as List<Object?>? ?? const [])
                .map((e) => MothEntitlement.fromJson(e as Map<String, Object?>))
                .toList(growable: false),
        subscriptions: (json['subscriptions'] as List<Object?>? ?? const [])
            .map(
              (e) => MothActiveSubscription.fromJson(e as Map<String, Object?>),
            )
            .toList(growable: false),
      );

  /// The entitlements the user currently holds. Empty means the free `none`
  /// tier — a valid, expected state, not an error.
  final List<MothEntitlement> activeEntitlements;

  /// The user's known subscriptions across stores.
  final List<MothActiveSubscription> subscriptions;

  /// Whether the user currently holds the entitlement [identifier] (e.g.
  /// `pro`). The single question app code should ask to gate a feature.
  bool hasEntitlement(String identifier) =>
      activeEntitlements.any((e) => e.identifier == identifier);

  /// The held entitlement with [identifier], or null.
  MothEntitlement? entitlement(String identifier) =>
      activeEntitlements.where((e) => e.identifier == identifier).firstOrNull;

  Map<String, Object?> toJson() => {
    'activeEntitlements': activeEntitlements
        .map((e) => e.toJson())
        .toList(growable: false),
    'subscriptions': subscriptions
        .map((s) => s.toJson())
        .toList(growable: false),
  };

  @override
  bool operator ==(Object other) =>
      other is MothCustomerInfo &&
      _listEquals(other.activeEntitlements, activeEntitlements) &&
      _listEquals(other.subscriptions, subscriptions);

  @override
  int get hashCode => Object.hash(
    Object.hashAll(activeEntitlements),
    Object.hashAll(subscriptions),
  );
}

bool _listEquals<T>(List<T> a, List<T> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i++) {
    if (a[i] != b[i]) return false;
  }
  return true;
}

File: sdk/react/src/core/customerInfo.ts
import { create, toBinary, fromBinary } from '@bufbuild/protobuf'
import { timestampDate, timestampFromDate } from '@bufbuild/protobuf/wkt'
import {
  CustomerInfoSchema,
  EntitlementSource,
  Store,
  SubscriptionStatus,
  type CustomerInfo,
} from '../gen/moth/billing/v1/billing_pb.js'

/** Which store a purchase or subscription belongs to. */
export type MothStore = 'apple' | 'google' | 'stripe'

/** Why an entitlement is active. */
export type MothEntitlementSource = 'store' | 'grant' | 'none'

/**
 * The store's renewal state, mapped to a small set common to every store.
 * active/trialing/inGracePeriod/inBillingRetry keep access;
 * paused/expired/revoked do not.
 */
export type MothSubscriptionStatus =
  | 'unspecified'
  | 'active'
  | 'trialing'
  | 'inGracePeriod'
  | 'inBillingRetry'
  | 'paused'
  | 'expired'
  | 'revoked'

/** Whether `status` keeps the subscription's access. */
export function subscriptionStatusIsActive(
  status: MothSubscriptionStatus,
): boolean {
  return (
    status === 'active' ||
    status === 'trialing' ||
    status === 'inGracePeriod' ||
    status === 'inBillingRetry'
  )
}

/**
 * One active capability the user holds (e.g. `pro`), with its expiry and why
 * it is active. Apps gate on {@link identifier}, never on a product id.
 */
export interface MothEntitlement {
  /** Stable identifier the app checks (e.g. `pro`). */
  identifier: string
  /** When the entitlement lapses; undefined for a non-expiring grant. */
  expireTime?: Date
  /** Why it is active (store subscription vs operator grant). */
  source: MothEntitlementSource
  /** The moth product identifier that granted it; empty for grants. */
  productIdentifier: string
}

/** One of the user's store subscriptions (may be inactive, for history). */
export interface MothActiveSubscription {
  productIdentifier: string
  /** The store this subscription lives on; undefined when unspecified. */
  store?: MothStore
  status: MothSubscriptionStatus
  /** End of the current paid (or trial) period. */
  currentPeriodEnd?: Date
  autoRenew: boolean
  /** Whether this subscription is a sandbox/test purchase. */
  isSandbox: boolean
}

/**
 * The complete subscription picture for one user, from
 * `moth.billing.v1.GetCustomerInfo`.
 *
 * A never-paid user, a free-tier user, and a user in a project with no
 * products all get a well-formed instance with empty entitlements (the
 * built-in `none` tier) — never an error. Gate with
 * {@link MothCustomerInfo.hasEntitlement}; never special-case "never paid".
 */
export class MothCustomerInfo {
  readonly activeEntitlements: readonly MothEntitlement[]
  readonly subscriptions: readonly MothActiveSubscription[]

  constructor(
    activeEntitlements: readonly MothEntitlement[] = [],
    subscriptions: readonly MothActiveSubscription[] = [],
  ) {
    this.activeEntitlements = activeEntitlements
    this.subscriptions = subscriptions
  }

  /** The valid, empty state: no entitlements, the free `none` tier. */
  static free(): MothCustomerInfo {
    return freeInfo
  }

  static fromProto(proto: CustomerInfo): MothCustomerInfo {
    return new MothCustomerInfo(
      proto.activeEntitlements.map((e) => {
        const ent: MothEntitlement = {
          identifier: e.identifier,
          source: sourceFromProto(e.source),
          productIdentifier: e.productIdentifier,
        }
        if (e.expireTime) ent.expireTime = timestampDate(e.expireTime)
        return ent
      }),
      proto.subscriptions.map((s) => {
        const sub: MothActiveSubscription = {
          productIdentifier: s.productIdentifier,
          status: statusFromProto(s.status),
          autoRenew: s.autoRenew,
          isSandbox: s.isSandbox,
        }
        const store = storeFromProto(s.store)
        if (store) sub.store = store
        if (s.currentPeriodEnd) {
          sub.currentPeriodEnd = timestampDate(s.currentPeriodEnd)
        }
        return sub
      }),
    )
  }

  /**
   * Whether the user currently holds the entitlement `identifier` (e.g.
   * `pro`) — the single question app code should ask to gate a feature. An
   * entitlement past its expiry no longer counts, so a cached snapshot
   * cannot grant lapsed access.
   */
  hasEntitlement(identifier: string, now: number = Date.now()): boolean {
    return this.entitlement(identifier, now) !== undefined
  }

  /** The held entitlement with `identifier`, or undefined. */
  entitlement(
    identifier: string,
    now: number = Date.now(),
  ): MothEntitlement | undefined {
    return this.activeEntitlements.find(
      (e) =>
        e.identifier === identifier &&
        (e.expireTime === undefined || e.expireTime.getTime() > now),
    )
  }

  equals(other: MothCustomerInfo): boolean {
    return (
      this.activeEntitlements.length === other.activeEntitlements.length &&
      this.subscriptions.length === other.subscriptions.length &&
      this.activeEntitlements.every((e, i) =>
        entitlementEquals(e, other.activeEntitlements[i]!),
      ) &&
      this.subscriptions.every((s, i) =>
        subscriptionEquals(s, other.subscriptions[i]!),
      )
    )
  }

  /**
   * Re-encodes this snapshot into the `moth.billing.v1.CustomerInfo` wire
   * schema — the payload the per-user cache persists (protobuf, never JSON).
   */
  toProtoBytes(): Uint8Array {
    return toBinary(
      CustomerInfoSchema,
      create(CustomerInfoSchema, {
        activeEntitlements: this.activeEntitlements.map((e) => ({
          identifier: e.identifier,
          source: sourceToProto(e.source),
          productIdentifier: e.productIdentifier,
          ...(e.expireTime ? { expireTime: timestampFromDate(e.expireTime) } : {}),
        })),
        subscriptions: this.subscriptions.map((s) => ({
          productIdentifier: s.productIdentifier,
          store: storeToProto(s.store),
          status: statusToProto(s.status),
          autoRenew: s.autoRenew,
          isSandbox: s.isSandbox,
          ...(s.currentPeriodEnd
            ? { currentPeriodEnd: timestampFromDate(s.currentPeriodEnd) }
            : {}),
        })),
      }),
    )
  }

  static fromProtoBytes(bytes: Uint8Array): MothCustomerInfo {
    return MothCustomerInfo.fromProto(fromBinary(CustomerInfoSchema, bytes))
  }
}

const freeInfo = new MothCustomerInfo()

function entitlementEquals(a: MothEntitlement, b: MothEntitlement): boolean {
  return (
    a.identifier === b.identifier &&
    a.source === b.source &&
    a.productIdentifier === b.productIdentifier &&
    (a.expireTime?.getTime() ?? -1) === (b.expireTime?.getTime() ?? -1)
  )
}

function subscriptionEquals(
  a: MothActiveSubscription,
  b: MothActiveSubscription,
): boolean {
  return (
    a.productIdentifier === b.productIdentifier &&
    a.store === b.store &&
    a.status === b.status &&
    a.autoRenew === b.autoRenew &&
    a.isSandbox === b.isSandbox &&
    (a.currentPeriodEnd?.getTime() ?? -1) === (b.currentPeriodEnd?.getTime() ?? -1)
  )
}

function sourceFromProto(source: EntitlementSource): MothEntitlementSource {
  switch (source) {
    case EntitlementSource.STORE:
      return 'store'
    case EntitlementSource.GRANT:
      return 'grant'
    default:
      return 'none'
  }
}

function sourceToProto(source: MothEntitlementSource): EntitlementSource {
  switch (source) {
    case 'store':
      return EntitlementSource.STORE
    case 'grant':
      return EntitlementSource.GRANT
    default:
      return EntitlementSource.NONE
  }
}

function storeFromProto(store: Store): MothStore | undefined {
  switch (store) {
    case Store.APPLE:
      return 'apple'
    case Store.GOOGLE:
      return 'google'
    case Store.STRIPE:
      return 'stripe'
    default:
      return undefined
  }
}

function storeToProto(store: MothStore | undefined): Store {
  switch (store) {
    case 'apple':
      return Store.APPLE
    case 'google':
      return Store.GOOGLE
    case 'stripe':
      return Store.STRIPE
    default:
      return Store.UNSPECIFIED
  }
}

function statusFromProto(status: SubscriptionStatus): MothSubscriptionStatus {
  switch (status) {
    case SubscriptionStatus.ACTIVE:
      return 'active'
    case SubscriptionStatus.TRIALING:
      return 'trialing'
    case SubscriptionStatus.IN_GRACE_PERIOD:
      return 'inGracePeriod'
    case SubscriptionStatus.IN_BILLING_RETRY:
      return 'inBillingRetry'
    case SubscriptionStatus.PAUSED:
      return 'paused'
    case SubscriptionStatus.EXPIRED:
      return 'expired'
    case SubscriptionStatus.REVOKED:
      return 'revoked'
    default:
      return 'unspecified'
  }
}

function statusToProto(status: MothSubscriptionStatus): SubscriptionStatus {
  switch (status) {
    case 'active':
      return SubscriptionStatus.ACTIVE
    case 'trialing':
      return SubscriptionStatus.TRIALING
    case 'inGracePeriod':
      return SubscriptionStatus.IN_GRACE_PERIOD
    case 'inBillingRetry':
      return SubscriptionStatus.IN_BILLING_RETRY
    case 'paused':
      return SubscriptionStatus.PAUSED
    case 'expired':
      return SubscriptionStatus.EXPIRED
    case 'revoked':
      return SubscriptionStatus.REVOKED
    default:
      return SubscriptionStatus.UNSPECIFIED
  }
}
